### PR TITLE
RFC 3161 Timestamping Support

### PR DESCRIPTION
**Pull request by mikhailswift**
* Adds an array to the signatures on a dsse envelope representing
  timestamps of that signature
* Adds implementation of RFC3161 timestamps for signing and policy
  enforcement
* Adds policy support for ensuring signatures have timestamps from a
  trusted time stamp authority

### CODE REVIEW

**Comment by mikhailswift** [comment]
I broke up the dsse sign and verify code into their separate files and it busted the diff :/. 

I'll try and re-arrange the commits so the file split happens separately, otherwise reviewing this is gonna be a pain.

File: dsse/dsse.go [hunk 1]
```diff
@@ -61,12 +54,20 @@ type Envelope struct {
...
 type Signature struct {
-	KeyID         string   `json:"keyid"`
-	Signature     []byte   `json:"sig"`
-	Certificate   []byte   `json:"certificate,omitempty"`
-	Intermediates [][]byte `json:"intermediates,omitempty"`
+	KeyID         string               `json:"keyid"`
+	Signature     []byte               `json:"sig"`
+	Certificate   []byte               `json:"certificate,omitempty"`
+	Intermediates [][]byte             `json:"intermediates,omitempty"`
+	Timestamps    []SignatureTimestamp `json:"timestamps,omitempty"`
```
Comment (colek42) [inline dsse/dsse.go:61]: This looks like a timestamp signature on the signature, correct?

File: dsse/dsse.go [hunk 2]
```diff
@@ -61,12 +54,20 @@ type Envelope struct {
...
+}
+
+type SignatureTimestampType string
+
+const TimestampRFC3161 SignatureTimestampType = "tsp"
 
-	trustedTime time.Time
+type SignatureTimestamp struct {
+	Type SignatureTimestampType `json:"type"`
+	Data []byte                 `json:"data"`
```
Comment (colek42) [inline dsse/dsse.go:70]: How is this data encoded?

File: dsse/sign.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+type SignOption func(*signOptions)
+
+func SignWithSigners(signers ...cryptoutil.Signer) SignOption {
+	return func(so *signOptions) {
+		so.signers = signers
+	}
+}
+
+func SignWithTimestampers(timestampers ...Timestamper) SignOption {
```
Comment (colek42) [inline dsse/sign.go:29]: I like how you implement this as an option

File: go.mod [hunk 4]
```diff
@@ -6,6 +6,7 @@ require (
 	github.com/anchore/stereoscope v0.0.0-20220708133445-777471f38c5b
 	github.com/anchore/syft v0.53.0
 	github.com/davecgh/go-spew v1.1.1
+	github.com/digitorus/timestamp v0.0.0-20220704143351-8225fba02d52
```
Comment (colek42) [inline go.mod:9]: Eventually, I'd like this to be in tree or a fork under our control

File: go.mod [hunk 5]
```diff
@@ -36,6 +37,7 @@ require (
 	github.com/containerd/containerd v1.6.6 // indirect
 	github.com/containerd/stargz-snapshotter/estargz v0.12.0 // indirect
 	github.com/coreos/go-oidc/v3 v3.2.0 // indirect
+	github.com/digitorus/pkcs7 v0.0.0-20220704143225-a9c8106cbfc6 // indirect
```
Comment (colek42) [inline go.mod:40]: See above comment

File: dsse/dsse.go (hunk 1 above)
Comment (mikhailswift) [reply dsse/dsse.go:61]: Correct

File: dsse/dsse.go (hunk 2 above)
Comment (mikhailswift) [reply dsse/dsse.go:70]: Base64. Go json marshals all byte arrays to base64.

File: go.mod (hunk 4 above)
Comment (mikhailswift) [reply go.mod:9]: Maintaining a fork of every lib we use is a maintenance nightmare

Comment (mikhailswift) [reply go.mod:9]: As far as what data actually is that is an implementation detail of the timestamper being used. 

In the case of TSP it’s an ASN1 encoded Timestamp Request object.

File: dsse/dsse.go (hunk 2 above)
Comment (mikhailswift) [reply dsse/dsse.go:70]: As far as what data actually is that is an implementation detail of the timestamper being used.

In the case of TSP it’s an ASN1 encoded Timestamp Request object.

File: go.mod (hunk 4 above)
Comment (mikhailswift) [reply go.mod:9]: We need to be judicial about what we fork and why. 

Forking repos incurs maintenance costs and burdens.

Comment (colek42) [reply go.mod:9]: I looked into it a bit more; the repo is maintained by a real company, https://www.digitorus.com/, based in the Netherlands that specializes in PKI.  I don't see any immediate concerns.

Initially, seeing a small repo with only two maintainers on a critical code path gives me pause.

File: timestamp/tsp.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+func TimestampWithRequestCertificate(requestCertificate bool) TSPTimestamperOption {
+	return func(t *TSPTimestamper) {
+		t.requestCertificate = requestCertificate
+	}
+}
+
+func NewTimestamper(opts ...TSPTimestamperOption) TSPTimestamper {
+	t := TSPTimestamper{
+		hash:               crypto.SHA256,
+		requestCertificate: true,
```
Comment (colek42) [inline timestamp/tsp.go:47]: Have you had any issues with the server not returning certs?  I think I remember somewhere that some TSAs do not implement this

Comment (mikhailswift) [reply timestamp/tsp.go:47]: Not currently, though I've only really tested with freetsa.org currently.

We may want to add some logic to check if there is no certificate and if so reject the timestamp perhaps

**Comment by colek42** [comment]
LGTM!  however, I think we need to make an issue for offline testing.

**Review by colek42 — APPROVED** [review]

**Pull request merged**